File: publication-figure-renderer/python/tests/test_svg_deidentifier.py
import json
from pathlib import Path
from typing import Any

import pytest
from lxml import etree

from publication_figure_renderer.config import LabelStyle
from publication_figure_renderer.label_mapper import build_label_mapping
from publication_figure_renderer.svg_deidentifier import (
    InvalidSvgError,
    deidentify_svg,
)

_FIXTURES = Path(__file__).resolve().parents[2] / "fixtures"
_SVG_NS = "http://www.w3.org/2000/svg"


def _sample_svg() -> str:
    return (_FIXTURES / "sample.svg").read_text(encoding="utf-8")


def _expected_deidentified_svg() -> str:
    return (_FIXTURES / "deidentified.svg").read_text(encoding="utf-8")


def _sample_detail() -> dict[str, Any]:
    payload: dict[str, Any] = json.loads(
        (_FIXTURES / "sample-detail.json").read_text(encoding="utf-8")
    )
    return payload


def _generation_mapping() -> dict[str, str]:
    detail = _sample_detail()
    id_to_label = build_label_mapping(detail, LabelStyle.GENERATION_NUMBER)
    individuals = detail["individuals"]
    return {ind["display_name"]: id_to_label[ind["id"]] for ind in individuals}


def _text_of(node: etree._Element) -> str:
    parts: list[str] = []
    if node.text is not None:
        parts.append(node.text)
    for child in node:
        parts.append(_text_of(child))
        if child.tail is not None:
            parts.append(child.tail)
    return "".join(parts)


def _text_contents(svg_text: str) -> list[str]:
    root = etree.fromstring(svg_text.encode("utf-8"))
    return [_text_of(t) for t in root.iter(f"{{{_SVG_NS}}}text")]


def _attribute_snapshot(svg_text: str, local_name: str) -> list[dict[str, str]]:
    root = etree.fromstring(svg_text.encode("utf-8"))
    return [_attrs_as_dict(node) for node in root.iter(f"{{{_SVG_NS}}}{local_name}")]


def _attrs_as_dict(node: etree._Element) -> dict[str, str]:
    out: dict[str, str] = {}
    for key, value in node.attrib.items():
        out[str(key)] = str(value)
    return out


def test_display_names_disappear_when_replaced_with_generation_number_labels() -> None:
    rendered = deidentify_svg(_sample_svg(), _generation_mapping())

    for name in (
        "Robert Smith",
        "Mary Smith",
        "David Smith",
        "Linda <O'Brien> & Co",
        "Sarah Smith",
    ):
        assert name not in rendered

    assert sorted(_text_contents(rendered)) == sorted(
        ["I-1", "I-2", "II-1", "II-2", "III-1"]
    )


def test_non_text_attributes_are_preserved_structurally() -> None:
    rendered = deidentify_svg(_sample_svg(), _generation_mapping())

    for name in ("rect", "circle", "polygon", "line"):
        assert _attribute_snapshot(rendered, name) == _attribute_snapshot(
            _sample_svg(), name
        ), name


def test_names_containing_xml_special_characters_are_handled_safely() -> None:
    rendered = deidentify_svg(_sample_svg(), {"Linda <O'Brien> & Co": "II-2"})

    assert "II-2" in rendered
    assert "Linda" not in rendered
    # Round-trip still parses as XML.
    etree.fromstring(rendered.encode("utf-8"))


def test_a_label_with_injection_shaped_characters_is_text_escaped_on_output() -> None:
    rendered = deidentify_svg(_sample_svg(), {"Robert Smith": "<script>alert(1)</script>"})

    assert "<script>" not in rendered
    assert "&lt;script&gt;" in rendered
    etree.fromstring(rendered.encode("utf-8"))


def test_empty_string_label_removes_the_matching_text_element() -> None:
    rendered = deidentify_svg(_sample_svg(), {"Sarah Smith": ""})

    contents = _text_contents(rendered)
    assert "Sarah Smith" not in contents
    assert "Robert Smith" in contents
    assert "Mary Smith" in contents


def test_width_and_height_overrides_update_root_only() -> None:
    rendered = deidentify_svg(_sample_svg(), {}, width=1024, height=768)

    root = etree.fromstring(rendered.encode("utf-8"))
    assert root.get("width") == "1024"
    assert root.get("height") == "768"
    original = etree.fromstring(_sample_svg().encode("utf-8"))
    assert root.get("viewBox") == original.get("viewBox")


def test_matches_the_canonical_deidentified_fixture_structurally() -> None:
    rendered = deidentify_svg(_sample_svg(), _generation_mapping())

    assert sorted(_text_contents(rendered)) == sorted(
        _text_contents(_expected_deidentified_svg())
    )
    rendered_root = etree.fromstring(rendered.encode("utf-8"))
    expected_root = etree.fromstring(_expected_deidentified_svg().encode("utf-8"))
    assert len(list(rendered_root.iter())) == len(list(expected_root.iter()))


def test_malformed_svg_raises_invalid_svg_error() -> None:
    with pytest.raises(InvalidSvgError):
        deidentify_svg("<svg><text>broken", {})

File: publication-figure-renderer/r/tests/testthat/test-svg_deidentifier.R
SVG_NS <- c(svg = "http://www.w3.org/2000/svg")

text_contents <- function(doc) {
  xml2::xml_text(xml2::xml_find_all(doc, "//svg:text", ns = SVG_NS))
}

attribute_snapshot <- function(doc, xpath) {
  nodes <- xml2::xml_find_all(doc, xpath, ns = SVG_NS)
  lapply(nodes, xml2::xml_attrs)
}

test_that("display names disappear when replaced with generation-number labels", {
  svg_text <- read_fixture_text("sample.svg")
  detail <- read_fixture_json("sample-detail.json")
  id_to_label <- build_label_mapping(detail, "generation-number")
  name_to_label <- stats::setNames(
    c(id_to_label[[detail$individuals[[1L]]$id]],
      id_to_label[[detail$individuals[[2L]]$id]],
      id_to_label[[detail$individuals[[3L]]$id]],
      id_to_label[[detail$individuals[[4L]]$id]],
      id_to_label[[detail$individuals[[5L]]$id]]),
    c("Robert Smith", "Mary Smith", "David Smith",
      "Linda <O'Brien> & Co", "Sarah Smith")
  )

  rendered <- deidentify_svg(svg_text, name_to_label)

  for (name in c("Robert Smith", "Mary Smith", "David Smith",
                 "Linda <O'Brien> & Co", "Sarah Smith")) {
    expect_false(grepl(name, rendered, fixed = TRUE))
  }

  rendered_doc <- xml2::read_xml(rendered)
  expect_equal(
    sort(text_contents(rendered_doc)),
    sort(c("I-1", "I-2", "II-1", "II-2", "III-1"))
  )
})

test_that("non-text attributes are preserved structurally", {
  svg_text <- read_fixture_text("sample.svg")
  original_doc <- xml2::read_xml(svg_text)
  detail <- read_fixture_json("sample-detail.json")
  id_to_label <- build_label_mapping(detail, "generation-number")
  name_to_label <- c(
    "Robert Smith" = unname(id_to_label[[detail$individuals[[1L]]$id]]),
    "Mary Smith" = unname(id_to_label[[detail$individuals[[2L]]$id]]),
    "David Smith" = unname(id_to_label[[detail$individuals[[3L]]$id]]),
    "Linda <O'Brien> & Co" = unname(id_to_label[[detail$individuals[[4L]]$id]]),
    "Sarah Smith" = unname(id_to_label[[detail$individuals[[5L]]$id]])
  )

  rendered_doc <- xml2::read_xml(deidentify_svg(svg_text, name_to_label))

  for (xpath in c("//svg:rect", "//svg:circle", "//svg:polygon", "//svg:line")) {
    expect_equal(
      attribute_snapshot(rendered_doc, xpath),
      attribute_snapshot(original_doc, xpath),
      info = xpath
    )
  }
})

test_that("names containing XML-special characters are handled safely", {
  svg_text <- read_fixture_text("sample.svg")
  name_to_label <- c("Linda <O'Brien> & Co" = "II-2")

  rendered <- deidentify_svg(svg_text, name_to_label)

  # Replacement succeeded.
  expect_true(grepl("II-2", rendered, fixed = TRUE))
  # The original, properly-escaped entity string is no longer present.
  expect_false(grepl("Linda", rendered, fixed = TRUE))
  # And the re-serialised document must still parse.
  expect_silent(xml2::read_xml(rendered))
})

test_that("a label containing injection-shaped characters is text-escaped on output", {
  svg_text <- read_fixture_text("sample.svg")
  name_to_label <- c("Robert Smith" = "<script>alert(1)</script>")

  rendered <- deidentify_svg(svg_text, name_to_label)

  expect_false(grepl("<script>", rendered, fixed = TRUE))
  expect_true(grepl("&lt;script&gt;", rendered, fixed = TRUE))
  # Round-trip must still parse.
  expect_silent(xml2::read_xml(rendered))
})

test_that("an empty-string label removes the <text> element rather than leaving an empty node", {
  svg_text <- read_fixture_text("sample.svg")
  name_to_label <- c("Sarah Smith" = "")

  rendered <- deidentify_svg(svg_text, name_to_label)

  rendered_doc <- xml2::read_xml(rendered)
  expect_false("Sarah Smith" %in% text_contents(rendered_doc))
  # All other labels survive.
  expect_true(all(c("Robert Smith", "Mary Smith") %in% text_contents(rendered_doc)))
})

test_that("width and height overrides update the root attributes only", {
  svg_text <- read_fixture_text("sample.svg")

  rendered <- deidentify_svg(svg_text, character(0), width = 1024L, height = 768L)

  rendered_doc <- xml2::read_xml(rendered)
  root <- xml2::xml_root(rendered_doc)
  expect_equal(xml2::xml_attr(root, "width"), "1024")
  expect_equal(xml2::xml_attr(root, "height"), "768")
  # viewBox must survive untouched -- it anchors the print-quality scaling.
  original_doc <- xml2::read_xml(svg_text)
  expect_equal(
    xml2::xml_attr(root, "viewBox"),
    xml2::xml_attr(xml2::xml_root(original_doc), "viewBox")
  )
})

test_that("matches the canonical deidentified fixture structurally", {
  svg_text <- read_fixture_text("sample.svg")
  expected_doc <- xml2::read_xml(read_fixture_text("deidentified.svg"))
  detail <- read_fixture_json("sample-detail.json")
  id_to_label <- build_label_mapping(detail, "generation-number")
  name_to_label <- c(
    "Robert Smith" = unname(id_to_label[[detail$individuals[[1L]]$id]]),
    "Mary Smith" = unname(id_to_label[[detail$individuals[[2L]]$id]]),
    "David Smith" = unname(id_to_label[[detail$individuals[[3L]]$id]]),
    "Linda <O'Brien> & Co" = unname(id_to_label[[detail$individuals[[4L]]$id]]),
    "Sarah Smith" = unname(id_to_label[[detail$individuals[[5L]]$id]])
  )

  rendered_doc <- xml2::read_xml(deidentify_svg(svg_text, name_to_label))

  expect_equal(
    sort(text_contents(rendered_doc)),
    sort(text_contents(expected_doc))
  )
  expect_equal(
    length(xml2::xml_find_all(rendered_doc, "//svg:*", ns = SVG_NS)),
    length(xml2::xml_find_all(expected_doc, "//svg:*", ns = SVG_NS))
  )
})

test_that("a malformed SVG raises invalid_svg_error", {
  expect_error(
    deidentify_svg("<svg><text>broken", character(0)),
    class = "invalid_svg_error"
  )
})
